### PR TITLE
Improve ctrl-c and sig INT catching in main UCM loop

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Currently, a Ctrl-c during a `.> run main` will crash all of ucm, which is a huge pain since ucm takes a while to load. The cloud team reported this was causing them some headaches.

This change gracefully catches and handles interrupts within `.> run main`.

## Implementation notes

Rewrites the global interrupt handling logic to allow gracefully overriding interrupt handling for sub-sections of UCM, while still ensuring we always have _some_ ctrl-c logic in place..

## Interesting/controversial decisions

Is our top-level interrupt handler actually any good? We could likely add some more specific handlers on certain commands (e.g. the individual `run` commands could probably use a custom handler).

This instrumentation may be useful for allowing Unison programs to catch interrupts on their own.

## Test coverage

I tested the following program:

```unison
main : '{IO, Exception} ()
main _ =
  !base.io.console.getLine
  -- hit ctrl-c without entering a line
  ()
```

```ucm
.> run main
^C
Aborted.
.>
```

I also tried it on infinite evaluation loops (which previously, a ctrl-c would quit UCM entirely)

E.g.

```unison
x : '{} List Nat
x _ = 1 +: !x

> !x
```

If you save this file, unison will enter an infinite evaluation loop, now if you mash ctrl-c it will bail to the prompt rather than quitting entirely. (Though it seems you might have to hit it a few times to get through).

## Loose ends

Seems self-contained to me. It'd be nice if we had a way to test or compile on windows.

### CODE REVIEW

File: unison-cli/src/Compat.hs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+{-# LANGUAGE CPP #-}
```
Comment (ChrisPenner) [inline unison-cli/src/Compat.hs:1]: Had to move this module into the library from the executable.

File: unison-cli/src/Compat.hs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+import Unison.Prelude
+import qualified UnliftIO
+
+#if defined(mingw32_HOST_OS)
+import qualified GHC.ConsoleHandler as WinSig
+#else
+import qualified System.Posix.Signals as Sig
+#endif
+
+defaultInterruptHandler :: IO (IO ())
```
Comment (ChrisPenner) [inline unison-cli/src/Compat.hs:19]: We can likely come up with a more helpful top-level handler at some point, but this is what we had before 🤷🏼‍♂️

**Review by pchiusano — APPROVED** [review]

**Comment by ceedubs** [comment]
@ChrisPenner I'm giving this a try right now and it doesn't kill ucm but my process also seems to still be running. I'll reach out to you on Slack.

**Comment by ChrisPenner** [comment]
@ceedubs  good catch, I think we might be using `forkIO` unsafely in a few spots which might mean async exceptions aren't being propagated as they should be. I'll dig in a bit.

**Comment by ChrisPenner** [comment]
I've opened an issue to discuss the fix for the runtime here: https://github.com/unisonweb/unison/pull/2739 .

That should probably be merged before this one.

**Comment by ChrisPenner** [comment]
#2739 is merged so this is ready to go as soon as @aryairani  or @mitchellwrosen  have a chance to look over it 👍🏼

**Review by mitchellwrosen — APPROVED** [review]

**Pull request merged**